src/kernel/pagecache.c

```diff
@@ -596,7 +600,8 @@ closure_function(1, 3, void, pagecache_write_sg,
         pn->length = q.end;
 
     /* prepare pages for writing */
-    merge m = allocate_merge(pc->h, closure(pc->h, pagecache_write_sg_finish, pn, q, sg, completion, false));
+    merge m = allocate_merge(pc->h, closure(pc->h, pagecache_write_sg_finish,
+        get_current_thread(), pn, q, sg, completion, false));
```

Comment:
This seems reasonable, but shouldn't such restoration happen on the read path as well?

Reply:
In the read path, SG buffers are allocated by the kernel, and the pagecache code never touches user buffers. It's the completion routine that should restore the thread before copying data to user buffers (as done e.g. by iov_read_complete() and file_read_complete()).

Reply:
Right, my mistake.